### PR TITLE
Support extra feature columns in input datasets

### PR DESCRIPTION
**Pull request by jxl26**
## Description
Add support for a --extra-feature-columns flag that allows the user to point to certain column(s) of the dataset that represent extra features (like temperature, pressure, etc.) that they want to augment to the training input.

## Example / Current workflow
The current workflow has no support extra input features - unassigned columns will all be treated as target columns.

## Bugfix / Desired workflow
The new flag now allows the user to explicitly identify the extra features, which would be augmented to the input data.

## Questions
N/A

## Relevant issues
#1238 : user asks for flexible incorporation of external conditional features

## Checklist
- [Yes] linted with flake8?
- [Yes] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by JacksonBurns** [comment]
@jxl26 thanks for getting this underway! Some larger organizational notes on this approach, since I think we need to expand the scope a bit:
 - we should refer to these additional numbers as 'descriptors'. This new flag is essentially an alternative to the `--descriptors-path` argument which allows the input descriptors to be in the input file, so the new flag should be something like `--descriptors-column`. The description should also look something more like the [`--descriptors-path` description](https://github.com/chemprop/chemprop/blob/2df18f477b1e254e55335868ff293c056e02b178/chemprop/cli/common.py#L109-L113).
 - The changes should go in `common.py` rather than `train.py` so that users can train a model with this flag _and_ make predictions, optimize, etc. with it as well

Let me know if this makes sense, and if you are willing to rework this!

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -557,6 +557,14 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
         help="Seed for PyTorch randomness (e.g., random initial weights)",
     )
 
+    parser.add_argument(
+        "--extra-feature-columns",
+        type=str,
+        nargs="+",
+        default=None,
```
Comment (KnathanM) [inline chemprop/cli/train.py:564]:
```suggestion
```
This is already the default I think.

File: chemprop/cli/train.py [hunk 2]
```diff
@@ -557,6 +557,14 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
...
+        help="Extra features, like temperature and pressure, to be included in datasets",
```
Comment (KnathanM) [inline chemprop/cli/train.py:565]:
```suggestion
        help="Extra datapoint descriptors, like temperature and pressure, to be included in datasets",
```

File: chemprop/cli/utils/parsing.py [hunk 3]
```diff
@@ -74,7 +78,7 @@ def parse_csv(
         lt_mask = None
         gt_mask = None
 
-    return smiss, rxnss, Y, weights, lt_mask, gt_mask
+    return smiss, rxnss, X_extra, Y, weights, lt_mask, gt_mask
```
Comment (KnathanM) [inline chemprop/cli/utils/parsing.py:81]: I think my preference is to put X_extra at the end of the list because it is more optional than Y. People might be using our CLI code (they shouldn't but might be) and do something like `smiss, rxnss, Y, *_ = parse_csv` which would still work if we put X_extra at the end.

File: tests/cli/test_cli_regression_mol.py [hunk 4]
```diff
@@ -18,6 +18,7 @@
 def data_path(data_dir):
     return (
         str(data_dir / "regression" / "mol" / "mol.csv"),
+        str(data_dir / "regression" / "mol" / "augmented_mol.csv"),
```
Comment (KnathanM) [inline tests/cli/test_cli_regression_mol.py:21]: It looks like you only use this in one test. In this case, I think it is better to make it its own fixture instead of changing the data_path fixture.

File: tests/data/regression/mol/augmented_mol.csv [hunk 5]
```diff
@@ -0,0 +1,7 @@
+smiles,y,temperature,pressure,split
+C,0.1,40,1,train
+CC,0.2,60,3,train
+CCC,0.3,80,5,train
+CCCC,0.4,95,7,val
+CCCCC,0.5,110,9,val
+CCCCCC,0.6,120,11,test
```
Comment (KnathanM) [inline tests/data/regression/mol/augmented_mol.csv:7]: I think our CLI requires two datapoints in the test set because of the optional batch normalization. We should probably change that so that if batch normalization isn't used, the dataloader doesn't drop_last.

File: chemprop/cli/utils/parsing.py (hunk 3 above)
Comment (jxl26) [reply chemprop/cli/utils/parsing.py:81]: Yea this reminds me that the sklearn modules I wrote use parse_csv as a helper, which should be fixed after this pull request go through.

File: tests/data/regression/mol/augmented_mol.csv (hunk 5 above)
Comment (jxl26) [reply tests/data/regression/mol/augmented_mol.csv:7]: Expanded test set to have 2 elements.

**Comment by jxl26** [comment]
> @jxl26 thanks for getting this underway! Some larger organizational notes on this approach, since I think we need to expand the scope a bit:
> 
> * we should refer to these additional numbers as 'descriptors'. This new flag is essentially an alternative to the `--descriptors-path` argument which allows the input descriptors to be in the input file, so the new flag should be something like `--descriptors-column`. The description should also look something more like the [`--descriptors-path` description](https://github.com/chemprop/chemprop/blob/2df18f477b1e254e55335868ff293c056e02b178/chemprop/cli/common.py#L109-L113).
> * The changes should go in `common.py` rather than `train.py` so that users can train a model with this flag _and_ make predictions, optimize, etc. with it as well
> 
> Let me know if this makes sense, and if you are willing to rework this!

This does make sense! I have reworked the PR to incorporate these suggestions.

File: chemprop/cli/utils/parsing.py [hunk 6]
```diff
@@ -85,7 +89,7 @@ def parse_csv(
         lt_mask = None
         gt_mask = None
 
-    return smiss, rxnss, Y, weights, lt_mask, gt_mask
+    return smiss, rxnss, Y, weights, lt_mask, gt_mask, X_extra
```
Comment (KnathanM) [inline chemprop/cli/utils/parsing.py:92]:
```suggestion
    return smiss, rxnss, Y, weights, lt_mask, gt_mask, X_d_extra
```

File: chemprop/cli/utils/parsing.py [hunk 7]
```diff
@@ -62,13 +63,16 @@ def parse_csv(
         rxnss = None
         input_cols = [df.columns[0]]
 
+    descriptor_cols = list(descriptor_cols or [])
+    X_extra = df[descriptor_cols].to_numpy(np.single) if descriptor_cols else None
```
Comment (KnathanM) [inline chemprop/cli/utils/parsing.py:67]:
```suggestion
    X_d_extra = df[descriptor_cols].to_numpy(np.single) if descriptor_cols else None
```

**Review by KnathanM — APPROVED** [review]
